src/box/txn.h

```diff
@@ -777,6 +780,52 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type);
 int
 txn_begin_in_engine(struct engine *engine, struct txn *txn);
 
+static inline bool
+txn_is_linearizable(const struct txn *txn)
+{
+	return txn->isolation == TXN_ISOLATION_LINEARIZABLE;
+}
+
+/**
+ * Check if a space supports linearizable access, i.e. it is either synchronous
+ * or not replicated at all, and its engine prevents dirty reads.
+ */
+static inline int
+txn_check_space_linearizability(const struct txn *txn,
+				const struct space *space)
+{
+	assert(txn_is_linearizable(txn));
```

Comment:
Since `txn_is_linearizable()` and `txn_check_space_linearizability()` are always called together, I think we should merge them and call the resulting function simply `txn_check_space()` or something like that. I don't think we need to provide a public `txn_is_linearizable()` helper, as we don't have similar helpers for other isolation levels. Also, it looks this function should be defined as static in `txn.c`.

Reply:
Ok, merged.
I don't want to make it local to `txn.c` though, because this way it'll stop being inlined to `txn_begin_ro_stmt`

<details><summary>diff</summary>

```diff
diff --git a/src/box/txn.c b/src/box/txn.c
index 2eb6046e2..289fc3971 100644
--- a/src/box/txn.c
+++ b/src/box/txn.c
@@ -609,10 +609,8 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type)
                txn_set_flags(txn, TXN_IS_CONFLICTED);
        }
 
-       if (space != NULL && txn_is_linearizable(txn) &&
-           txn_check_space_linearizability(txn, space) != 0) {
+       if (space != NULL && txn_check_space_linearizability(txn, space) != 0)
                return -1;
-       }
 
        if (txn_check_can_continue(txn) != 0)
                return -1;
diff --git a/src/box/txn.h b/src/box/txn.h
index 2a2fd053d..ff84d2acc 100644
--- a/src/box/txn.h
+++ b/src/box/txn.h
@@ -780,12 +780,6 @@ txn_begin_stmt(struct txn *txn, struct space *space, uint16_t type);
 int
 txn_begin_in_engine(struct engine *engine, struct txn *txn);
 
-static inline bool
-txn_is_linearizable(const struct txn *txn)
-{
-       return txn->isolation == TXN_ISOLATION_LINEARIZABLE;
-}
-
 /**
  * Check if a space supports linearizable access, i.e. it is either synchronous
  * or not replicated at all, and its engine prevents dirty reads.
@@ -794,8 +788,8 @@ static inline int
 txn_check_space_linearizability(const struct txn *txn,
                                const struct space *space)
 {
-       assert(txn_is_linearizable(txn));
-       (void)txn;
+       if (txn->isolation != TXN_ISOLATION_LINEARIZABLE)
+               return 0;
        /*
         * XXX: can't promise the user linearizability with Vinyl until
         * https://github.com/tarantool/tarantool/issues/5522 is fixed.
@@ -845,10 +839,8 @@ txn_begin_ro_stmt(struct space *space, struct txn **txn,
                        diag_set(ClientError, ER_TRANSACTION_CONFLICT);
                        return -1;
                }
-               if (txn_is_linearizable(*txn) &&
-                   txn_check_space_linearizability(*txn, space) != 0) {
+               if (txn_check_space_linearizability(*txn, space) != 0)
                        return -1;
-               }
                struct engine *engine = space->engine;
                return txn_begin_in_engine(engine, *txn);
        }

```
</details>

Reply:
Nit: I'd call the new function simply `txn_check_space` (check that a space can be used in a transaction) so that we can use it for other stuff, too. Maybe, even do some refactoring like this:

```C
/** Checks that a space can be used in a transaction. */
void txn_check_space(struct txn *txn, struct space *space)
{
    if (txn->isolation == TXN_ISOLATION_LINEARIZABLE)
        txn_check_space_linearizability(space);
    /* We could place other checks here without patching the call sites. */
}
```